# EDA

In [1]:
import numpy as np
import pandas as pd
from sklearn.cluster import DBSCAN
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import StandardScaler

In [2]:
df=pd.read_csv('c:/users/gaurav yadav/downloads/crime_data.csv')
print(df.shape)
df.isna().sum()


(50, 5)


Unnamed: 0    0
Murder        0
Assault       0
UrbanPop      0
Rape          0
dtype: int64

In [3]:
#no null values so no need of imputation
#as states column doesn't convey much info so it is not considered in model building

In [4]:
x=df.iloc[:,1:].values
#FEATURE SCALING
sc_x=StandardScaler()
x=sc_x.fit_transform(x)


In [65]:
#model building
model=DBSCAN(eps=1.1,min_samples=3)
model.fit(x)


DBSCAN(eps=1.1, min_samples=3)

In [66]:
model.labels_

array([ 0, -1,  1,  2, -1, -1,  2, -1, -1,  0,  2,  2,  1,  2,  2,  2,  2,
        0,  2,  1,  2,  1,  2,  0,  2,  2,  2, -1,  2,  2,  1,  1,  0,  2,
        2,  2,  2,  2,  2,  0,  2,  0,  1,  2,  2,  2,  2,  2,  2,  2],
      dtype=int64)

In [67]:
df['cluster']=model.labels_


In [68]:
df.cluster.value_counts()

 2    30
 1     7
 0     7
-1     6
Name: cluster, dtype: int64

In [69]:
df1=df.groupby('cluster').agg('mean')

In [70]:
df1.sort_values('cluster')

Murder     Assault   UrbanPop       Rape
cluster                                             
-1       10.066667  261.333333  75.000000  36.250000
 0       14.671429  251.285714  54.285714  21.685714
 1       11.014286  262.571429  77.142857  28.800000
 2        4.973333  112.433333  63.566667  16.356667

In [93]:
def range_x(x):
    return (np.min(x),np.max(x))
for i in range(1,(df1.shape[1])+1):
    print(df.columns[i],range_x(df.iloc[:,i:i+1][df.cluster==2]))
    print()

Murder (Murder    0.8
dtype: float64, Murder    9.7
dtype: float64)

Assault (Assault    45
dtype: int64, Assault    190
dtype: int64)

UrbanPop (UrbanPop    32
dtype: int64, UrbanPop    89
dtype: int64)

Rape (Rape    7.3
dtype: float64, Rape    29.3
dtype: float64)



# eps value and min_samples-
'''as we reduce eps value number of outliers increases , and as we increases min_samples 
beyond 5 no. of outliers in data increases drastically,so optimum value of eps=1.1,and min_samples=3 we get the best result although these are not best combination ,bcz in one cluster we have 30 data points which is 60% of totol data points
in dataset but noise data points are in limit that is about 14% of total data points in dataset.


we should not use DBSCAN for this data set,we should use some other algo like KMeans or hierarchical clustering.'''

# Insights from clusters--'''the states which are treated as outliers are having maximum avg assault,avg urbanpop, avg rape and high avg murder may be that's why they are treated as outliers.                                                                    Cluster with minimum avg murder , avg assault and avg rape ,reason for that is bcz it contains maximum number of data points ,bcz of that variation is more.                                                                                                                          For remaining clusters there are less number of data points and variation with in a feature is very less .'''






# '''for cluster  with maximum data points cluster 2 variation of data with in a column. for murder column max =9.7 and min=.8                                                  for assault column  max=190,   min=45                                                                for rape column max=29.3 ,min=7.3'''

